File: src/main/java/me/viniciusarnhold/altaria/command/AbstractCommand.java
package me.viniciusarnhold.altaria.command;

import me.viniciusarnhold.altaria.command.interfaces.ICommand;
import me.viniciusarnhold.altaria.utils.Actions;
import org.jetbrains.annotations.NotNull;
import sx.blah.discord.api.events.IListener;
import sx.blah.discord.handle.impl.events.MessageReceivedEvent;
import sx.blah.discord.handle.obj.IMessage;
import sx.blah.discord.util.MessageBuilder;
import sx.blah.discord.util.RequestBuffer;

import java.util.Collections;
import java.util.EnumSet;
import java.util.Set;

/**
 * Created by Vinicius.
 *
 * @since ${PROJECT_VERSION}
 */
public abstract class AbstractCommand implements ICommand, IListener<MessageReceivedEvent> {

    protected String command;

    protected Set<String> aliases;

    protected String description;

    protected CommandType commandType;

    protected EnumSet<UserPermissions> permissions;

    public AbstractCommand() {
        command = "";
        aliases = Collections.emptySet();
        description = "";
        commandType = CommandType.GENERAL;
        permissions = EnumSet.noneOf(UserPermissions.class);
    }

    @NotNull
    public static RequestBuffer.RequestFuture<IMessage> enqueue(@NotNull MessageBuilder builder) {
        return RequestBuffer.request(Actions.wrap(builder::send));
    }

    @NotNull
    @Override
    public String command() {
        return command;
    }

    @NotNull
    @Override
    public Set<String> aliases() {
        return aliases;
    }

    @NotNull
    @Override
    public String description() {
        return description;
    }

    @NotNull
    @Override
    public CommandType type() {
        return commandType;
    }

    @NotNull
    @Override
    public EnumSet<UserPermissions> permissions() {
        return permissions;
    }

    protected boolean isMyCommand(@NotNull MessageReceivedEvent event) {
        return !event.getMessage().getChannel().isPrivate() &&
                !event.getMessage().getAuthor().isBot() &&
                MessageUtils.isMyCommand(event.getMessage(), this);
    }
}

File: core/src/main/kotlin/me/viniciusarnhold/altaria/command/AbstractCommand.kt
package me.viniciusarnhold.altaria.command

import me.viniciusarnhold.altaria.command.interfaces.ICommand
import me.viniciusarnhold.altaria.utils.Actions
import sx.blah.discord.api.events.IListener
import sx.blah.discord.handle.impl.events.MessageReceivedEvent
import sx.blah.discord.handle.obj.IMessage
import sx.blah.discord.util.MessageBuilder
import sx.blah.discord.util.RequestBuffer
import java.util.*

/**
 * Created by Vinicius.

 * @since ${PROJECT_VERSION}
 */
abstract class AbstractCommand : ICommand, IListener<MessageReceivedEvent> {

    protected var command: String

    protected var aliases: Set<String>

    protected var description: String

    protected var commandType: CommandType

    protected var permissions: EnumSet<UserPermissions>

    init {
        command = ""
        aliases = emptySet<String>()
        description = ""
        commandType = CommandType.GENERAL
        permissions = EnumSet.noneOf<UserPermissions>(UserPermissions::class.java)
    }

    override fun command(): String {
        return command
    }

    override fun aliases(): Set<String> {
        return aliases
    }

    override fun description(): String {
        return description
    }

    override fun type(): CommandType {
        return commandType
    }

    override fun permissions(): EnumSet<UserPermissions> {
        return permissions
    }

    protected fun isMyCommand(event: MessageReceivedEvent): Boolean {
        return !event.message.channel.isPrivate &&
                !event.message.author.isBot &&
                MessageUtils.isMyCommand(event.message, this)
    }

    companion object {

        fun enqueue(builder: MessageBuilder): RequestBuffer.RequestFuture<IMessage> {
            return RequestBuffer.request(Actions.wrap<IMessage>({ builder.send() }))
        }
    }
}
